### PR TITLE
Update/1702 restructure

### PR DESCRIPTION
**Pull request by mands**
Large amount of changes and restructuring for 17.02 launch.

### CODE REVIEW

File: features/index.rst [hunk 1]
```diff
@@ -1,14 +1,16 @@
-.. _workflow_power: 
+.. _features
 
-More Powerful Workflows
-=======================
+Features
+========
 
-Now that we have published our classifier to NStack as a module, we can use it to demonstrate some of the more powerful features of the workflow engine.
+In this section, we're going to describe some of the more advavcned featreus you can do with NStack when building your modules and composing them together to build workflows.
```
Comment (lanthias) [inline features/index.rst]: spelling

File: features/index.rst [hunk 2]
```diff
@@ -114,3 +118,68 @@ Because ``roundedPetalsSource`` is a combination of a source and a function, it
...
+        return full_name
+
+
+.. _features_framework
+
+
+Framework Modules
+-----------------
+
+You may want to create a common parent module that has lots of complex dependencies already installed, either to save time or for standardisation. NStack supports this with *Framework Modules*. Simply create a new module similar to above, ``nstack init framework [parent]``, and modify the resulting ``nstack.yaml`` as needed.
```
Comment (lanthias) [inline features/index.rst]: "It is often useful to create a common parent module with dependencies already installed"

File: features/index.rst [hunk 3]
```diff
@@ -114,3 +118,68 @@ Because ``roundedPetalsSource`` is a combination of a source and a function, it
...
+
+You can then build this module using ``nstack build``, and refer to it within your future modules within the ``parent`` field of their ``nstack.yaml`` config file.
```
Comment (lanthias) [inline features/index.rst]: files (plural)

File: languages.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
+.. _languages
+
+
+Supported Languages
+===================
+
+NStack allows you to write modules in your language of choice, currently we support Python, with R and Java coming soon. More languages will be added over time - if there's something you really need, please let us know!
```
Comment (lanthias) [inline languages.rst]: Maybe just say 'is language-agnostic'

File: languages.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  class Service(nstack.BaseService):
+      def numChars(self, msg):
+          return len(msg)
+
+.. note:: Ensure you import the nstack module in your service, e.g. ``import nstack`` 
+
+Any function that you export within the ``API`` section of your ``nstack yaml`` must exist as a method on this class (you can add private methods on this class for internal use as expected in Python).
+
+Data comes into this function via the method arguments - for ``nstack`` all the data is passed wiithin a single argument that follows the ``self`` parameter. For instance, above there is a single arguent ``msg`` that is a single ``string`` element to use directly. However if defined your function as taking ``(Double, Double)``, you would need to unpack the tuple in Python as follows, ::
```
Comment (lanthias) [inline languages.rst]: wiithin -> within

File: languages.rst [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  def test(self, msg):
+      x, y = msg
+      ...
+
+Similarly, to return data from your NStack function, simply return it as a single element within your Python method, as in the top example above.
+
+The NStack object lasts for the life-time of the workflow, so if there is any initialisation you need to do within your service you can perform this within the object ``__init__`` method, e.g. open a database connection, load a data-set.
+However rememebr to call the parent object ``__init__`` method to ensure NStack is initialised correctly, i.e.. ::
```
Comment (lanthias) [inline languages.rst]: remember

File: languages.rst [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  class Service(nstack.BaseService):
+      def __init__(self):
+          super().__init__()
+          # custom initialisation here
+
+
+Notes
+^^^^^
+
+* Anything your ``print`` will show up in ``nstack log`` to aid debugging. (all output on ``stdout`` and ``stderr`` is sent to the NStack logs)
```
Comment (lanthias) [inline languages.rst]: your -> you

File: reference/nstack_toolkit.rst [hunk 8]
```diff
@@ -1,13 +1,11 @@
 .. _nstack_cli:
 
 nstack CLI
-================
+==========
 
+The nstack CLI is used to build modules and workflows on a linked NStack server.
+It can be configured via the ``nstack.conf`` ``yaml`` file found in ``~\.config`` on Linux/macOS, and in ``HOME/AppUser/Roaming`` on Windows, or indirectly via the ``nstack register`` command described below.
```
Comment (lanthias) [inline reference/nstack_toolkit.rst]: Do you mean set-server?

File: reference/nstack_toolkit.rst [hunk 9]
```diff
@@ -30,6 +28,48 @@ Commands
...
+=============    ===========
+Option           Description
+=============    ===========
+``username``     A unique username to assign on the the server.
+``email``        An email address to validate the user and send login credentials to.
+``SERVER``       The remote NStack Server to register with, by default this will use our demo server.
+=============    ===========
+
+A simple command to register with a remote NStack server so you can login, build modules, start workflows, etc.
+Upon sucessful regsitratio nyou will recevier credtiions in an email that you can paste into the ``nstack`` CLI and get started.
```
Comment (lanthias) [inline reference/nstack_toolkit.rst]: typos

File: reference/nstack_toolkit.rst [hunk 10]
```diff
@@ -30,6 +28,48 @@ Commands
...
+=============    ===========
+Option           Description
+=============    ===========
+``server-addr``  URL of a remote NStack Server.
+``server-port``  Port of a remote NStack Server.
+``id``           Your unique id used to communicate with the remote NStack Server.
+``key``          Your secret key used to communicate with the remote NStack Server
+=============    ===========
+
+This command configures your local NStack CLI to communicate with a remote NStack Server with which you have registered (see previous command). You usually don't have enter this command by hand, it will be contained with an email after sucessful registration that you can paste directly into your terminal.
```
Comment (lanthias) [inline reference/nstack_toolkit.rst]: "have to"

File: reference/nstack_toolkit.rst [hunk 11]
```diff
@@ -107,8 +147,49 @@ Or, if you have built a workflow as a module, you can start it with:
...
+``notebook`` 
+^^^^^^^^^^^^
+
+.. code:: bash
+
+    $ nstack notebook
+   
+Create an interactive session within the terminal that provides a mini-REPL (you can also redirect a file/stream into the notebook command to provide for rapid service testing and development). 
+
+From this command-line, you can import modules as needed, and enter a single workflow that wil lbe compiled and run immediately on the server (press ``<Ctrl-D>`` on Linux/macOS or ``<Ctrl-Z>`` on Windows to submit your input).
```
Comment (lanthias) [inline reference/nstack_toolkit.rst]: "will be"

File: reference/nstack_toolkit.rst [hunk 12]
```diff
@@ -107,8 +147,49 @@ Or, if you have built a workflow as a module, you can start it with:
...
+    $ nstack send "route" 'data'
+
+=============    ===========
+Option           Description
+=============    ===========
+``route``        The endpoint to send the data where a workflow is running.
+``data``         A json snippet to send to the endpoint and pass into a workflow.
+=============    ===========
+
+Used with the HTTP source, ``nstack send`` sends a JSON-encoded element to an endpoint on the NStack server where a workflow has been started. Usfule for testing workflows that are to be used as web-hooks.
```
Comment (lanthias) [inline reference/nstack_toolkit.rst]: "Useful"

File: features/index.rst (hunk 3 above)
Comment (mands) [reply features/index.rst]: No - singular.

**Review by lanthias — APPROVED** [review]

**Pull request merged**